### ISSUE TITLE
How to decode torch_generator_agent prediction vector back to text

### ISSUE THREAD

**Issue by rguan1** [facebookresearch/ParlAI#4431]
Hi,

I've been looking at the compute_loss() function in torch_generator_agent.py so that I can override this loss function. For part of my custom loss function, I want to decode the prediction vector back to text so that I can use that text to help compute my new loss function. However, I am not sure if I am decoding the prediction vector properly, because the text seems incoherent. So, it'd be very helpful to know how to properly decode prediction vectors back to text. I will describe my current process of decoding the text below. 

In compute_loss(), the model makes its predictions and then unpacks the prediction vector:
```
        model_output = self.model(*self._model_input(batch), ys=batch.label_vec)
        scores, preds, *_ = model_output
```

Then, I try to convert the vector `preds` back to text by invoking:
```
        text = [self._v2t(p) for p in preds]
```
The resulting `text` list contains these decoded predictions that seem incorrect: 
```
["i know . i was ' t imagine missing missing that situation . i been i nightmare . .", "i , , i ' bad i bob be ha been me . as he he is her girl f", "why ' . i are so many good people out hurt .", 'i , sure . i . i you have they can have have to your by ? the found in parent ?', 'i know , i guess understand why point " " we you don \' t do it , it for to get it die it " thing dset , but world to \' be is make sing the cities is but , towns . make full. just tough that', 'yeah know have it sad when i is is', 'i , i i ats even more sad is that fact of a on the government government .', "i ' s what , i , found not same thing by her her to a school kuwork .", 'i too . i to to the a good time is then a around it to you . don is', 'i . people be done to help it from', 'i think very about much could about the new in i sure the people people .', 'i , is true . i am it is to but i police is her .', 'i know , i a you think of happen done to help the', 'i is a interesting interesting . i kids kids .', "i , i then is is ' t enough .", 'i know have to agree with but a sad to believe . kind']
```

**Comment by stephenroller** [comment]
`self._v2t` is right.

**Comment by klshuster** [comment]
note that each of those decoded tokens is individually conditioned on the _teacher forced context_, so they might make sense given the gold tokens prior, but not all put together

**Comment by rguan1** [comment]
Is there a function that I can use to decode the actual utterances that blender would normally return like when I invoke act() on blender rather than decoding the individually conditioned tokens? I have been unable to find such a function / example in the repository somewhere so far. 

Thank you again for help!

**Comment by klshuster** [comment]
[here's the loss function for our unlikelihood agent](https://github.com/facebookresearch/ParlAI/blob/b6a51af905564f24b4bd6b22a78df892a36a6eae/projects/dialogue_unlikelihood/agents.py#L167). this loss function computes loss as usual, and then generates from the model (with a call to `self._generate`), and uses that generation for computing additional loss

**Comment by rguan1** [comment]
This is fantastic. Thank you so much! 

The `self._generate` call works when used in conjunction with self._v2t. Here's a snippet of code that decodes the training utterances for any folks that are facing the same issue in the future. The `text` var contains the decoded utterances.

```
      clamp_min = 1e-6 if self.opt['fp16'] else 1e-20
      maxlen = self.label_truncate or 256
      with torch.no_grad():
          beam_pred_scores, rest = self._generate(batch, self.beam_size, maxlen)

      generations = [g[1:] for (g, s, _) in beam_pred_scores]
      pred_toks = torch.nn.utils.rnn.pad_sequence(generations, batch_first=True)
      text = [self._v2t(p) for p in pred_toks]
```

**Comment by klshuster** [comment]
Great, closing for now, please reopen if there are further issues.
